storage/src/main/java/tech/pegasys/teku/storage/store/StoreTransactionUpdatesFactory.java

```diff
@@ -159,10 +170,37 @@ private StoreTransactionUpdates buildFinalizedUpdates(final Checkpoint finalized
     return states;
   }
 
-  private void calculatePrunedHotBlockRoots(StoreTransaction tx, final BlockTree prunedTree) {
-    tx.getBlockRoots().stream()
-        .filter(root -> !prunedTree.contains(root))
-        .forEach(prunedHotBlockRoots::add);
+  private boolean shouldPrune(
+      final SignedBeaconBlock finalizedBlock,
+      final Bytes32 blockRoot,
+      final UInt64 slot,
+      final Bytes32 parentRoot) {
+    return (slot.isLessThanOrEqualTo(finalizedBlock.getSlot())
+            || prunedHotBlockRoots.contains(parentRoot))
+        // Keep the actual finalized block
+        && !blockRoot.equals(finalizedBlock.getRoot());
+  }
+
+  private void calculatePrunedHotBlockRoots() {
+    final SignedBeaconBlock finalizedBlock = tx.getLatestFinalized().getBlock();
+    tx.blockAndStates.values().stream()
+        // Iterate new blocks in slot order to guarantee we see parents first
+        .sorted(Comparator.comparing(SignedBlockAndState::getSlot))
+        .filter(
+            newBlockAndState ->
+                shouldPrune(
+                    finalizedBlock,
+                    newBlockAndState.getRoot(),
+                    newBlockAndState.getSlot(),
+                    newBlockAndState.getParentRoot()))
+        .forEach(newBlockAndState -> prunedHotBlockRoots.add(newBlockAndState.getRoot()));
+
+    baseStore.blockMetadata.processAllInOrder(
```

Comment:
I think you need to process the base blocks first so the parentRoot-based filtering works

Reply:
ooo, good spot.

Reply:
We should add a test that covers this if we don't already have one.

Reply:
Done.